### PR TITLE
feature: eth address derivation from ecdsa pubkey

### PR DESCRIPTION
**Pull request by lucasmenendez**

### CODE REVIEW

File: address/address.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	// convert public key coords to uint8 and concatenate them
+	xBytes, err := elemToU8(api, pubKey.X)
+	if err != nil {
+		return 0, err
+	}
+	yBytes, err := elemToU8(api, pubKey.Y)
+	if err != nil {
+		return 0, err
+	}
+	pubBytes := append(xBytes, yBytes...)
```
Comment (p4u) [inline address/address.go:62]: Umm. Would it be an option to input an already reversed pubkey bytes to the circuit? If so, how many constraints do we reduce?

File: address/address_test.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		pubBytes = append(pubBytes, b)
+	}
+	for _, b := range privKey.PublicKey.A.Y.Bytes() {
+		pubBytes = append(pubBytes, b)
+	}
+	hash := keccak256.Hash(pubBytes)
+	// compute the signature of an arbitrary message
+	sigBin, err := privKey.Sign(input, nil)
+	if err != nil {
+		t.Fatal(err)
```
Comment (p4u) [inline address/address_test.go:54]: Use quicktest

File: address/address_test.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	r, s := new(big.Int), new(big.Int)
+	r.SetBytes(sig.R[:32])
+	s.SetBytes(sig.S[:32])
+	// get the address from the hash of the public key (taking the last 20 bytes
+	// of the Keccak-256 hash of the public key)
+	address := new(big.Int).SetBytes(hash[12:])
+	// compile the circuit and get the constraints
+	p := profile.Start()
+	now := time.Now()
+	_, _ = frontend.Compile(ecc.BLS12_377.ScalarField(), r1cs.NewBuilder, &testAddressCircuit{})
```
Comment (p4u) [inline address/address_test.go:72]: I don't think we need to compile the circuit for this test.

File: address/address_test.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	var sig ecdsa.Signature
+	if _, err := sig.SetBytes(sigBin); err != nil {
+		t.Fatal(err)
+	}
+	r, s := new(big.Int), new(big.Int)
+	r.SetBytes(sig.R[:32])
+	s.SetBytes(sig.S[:32])
+	// get the address from the hash of the public key (taking the last 20 bytes
+	// of the Keccak-256 hash of the public key)
+	address := new(big.Int).SetBytes(hash[12:])
```
Comment (p4u) [inline address/address_test.go:68]: I think it's better if we use go-ethereum common package to build the address

**Review by p4u — CHANGES_REQUESTED** [review]

File: address/address_test.go (hunk 3 above)
Comment (lucasmenendez) [reply address/address_test.go:72]: We need it if we want to know the number of constrains.

File: address/address.go (hunk 1 above)
Comment (p4u) [reply address/address.go:62]: Then we might use the Ethereum Address in big Endian format always :P

Joking, let's leave it as it is now. We'll think about optimizations in the future.

**Pull request merged**